build.rs writes to wrong directory

In general, build outputs should only be written to OUT_DIR.
(https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script, "Scripts should not modify any files outside of that directory.")

The existing `build.rs` also writes to CARGO_MANIFEST_DIR when the feature `assemble` is active: https://github.com/rp-rs/rp2040-boot2/blob/4c477f9e1a4ef0beff10bd3c4436f4e66fec334c/build.rs#L107-L111

While unusual, this is handy when called locally to update the pre-built binaries.

However, it becomes a problem when used as part of some dependency tree. For example, I have some firmware which depends on a local checkout of  `rp2040-boot` like this:
```
[dependency]
rp2040-boot2 = { version = "0.2", path = "/some/other/path/rp2040-boot2-rs", features=["assemble"] }
```

In this case, the build will write to `/some/other/path/rp2040-boot2-rs/bin`. This is unexpected, and might not even work. In my case, that path was not writable by the current build process, resulting in this error message:

```
error:failed to run custom build command for `rp2040-boot2 v0.2.2 (/some/other/path/rp2040-boot2-rs)`

Causedby:
 process didn't exit successfully: `/some/other/path/rp2040-project-template/target/debug/build/rp2040-boot2-ab9f999e2371819e/build-script-build` (exit status: 101)
 --- stdout
 cargo:warning=final out dir = /some/other/path/rp2040-boot2-rs/bin/

 --- stderr
 In file included from src/include/boot2_helpers/exit_from_boot2.S:10:0,
                  from src/boot2_at25sf128a.S:273:
 src/include/hardware/regs/m0plus.h:565:0: warning: "M0PLUS_CPUID_OFFSET" redefined
  #define M0PLUS_CPUID_OFFSET _u(0x0000ed00)
   
 In file included from src/include/pico/platform.h:20:0,
                  from src/include/pico.h:31,
                  from src/include/pico/asm_helper.S:7,
                  from src/boot2_at25sf128a.S:28:
 src/include/hardware/regs/addressmap.h:72:0: note: this is the location of the previous definition
  #define M0PLUS_CPUID_OFFSET 0x0000ed00
   
 In file included from src/include/boot2_helpers/exit_from_boot2.S:10:0,
                  from src/boot2_at25sf128a.S:273:
 src/include/hardware/regs/m0plus.h:742:0: warning: "M0PLUS_VTOR_OFFSET" redefined
  #define M0PLUS_VTOR_OFFSET _u(0x0000ed08)
   
 In file included from src/include/pico/platform.h:20:0,
                  from src/include/pico.h:31,
                  from src/include/pico/asm_helper.S:7,
                  from src/boot2_at25sf128a.S:28:
 src/include/hardware/regs/addressmap.h:73:0: note: this is the location of the previous definition
  #define M0PLUS_VTOR_OFFSET 0x0000ed08
   
 /usr/lib/gcc/arm-none-eabi/7.3.1/../../../arm-none-eabi/bin/ld: warning: cannot find entry symbol _start; defaulting to 0000000000008000
 thread 'main' panicked at 'writing padded output file: Os { code: 30, kind: ReadOnlyFilesystem, message: "Read-only file system" }', /some/other/path/rp2040-boot2-rs/build.rs:99:34
 note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

When not using a `path` dependency, it writes to paths like `~/.cargo/registry/src/github.com-1ecc6299db9ec823/rp2040-boot2-0.2.1/bin/boot2_at25sf128a.padded.bin` instead, which isn't correct either.

--- a/README.md
+++ b/README.md
@@ -72,7 +72,7 @@ In order to remove the need for GCC for users of this crate, we link against pre
 If you wish to add or change an existing bootloader you should install GCC and build with the feature `assemble`
 
 ```
-cargo build --features=assemble
+UPDATE_PRECOMPILED_BINARIES=true cargo build --features=assemble
 ```
 
 To add a new bootloader to the build you need to add it to `SOURCE_FILES` in `build.rs` and add an entry for it in `lib.rs`

--- a/build.rs
+++ b/build.rs
@@ -1,14 +1,13 @@
 //! Compiles boot2 bootloader from assembler source
 //! Compiles the bootloader from assembly language source, and creates a binary file.
 
-#[cfg(feature = "assemble")]
 use std::env;
-#[cfg(feature = "assemble")]
 use std::fs;
 #[cfg(feature = "assemble")]
 use std::io::{self, Write};
+use std::path::Path;
 #[cfg(feature = "assemble")]
-use std::path::{Path, PathBuf};
+use std::path::PathBuf;
 #[cfg(feature = "assemble")]
 use std::process::Command;
 
@@ -70,7 +69,7 @@ fn make_bin<P: AsRef<Path>, Q: AsRef<Path>>(input_path: P, out_dir: Q) -> PathBu
 }
 
 #[cfg(feature = "assemble")]
-fn make_padded_bin<P: AsRef<Path>, Q: AsRef<Path>>(input_path: P, out_dir: Q) {
+fn make_padded_bin<P: AsRef<Path>, Q: AsRef<Path>>(input_path: P, out_dir: Q) -> PathBuf {
     const BOOT2_OUTPUT_LEN: usize = 256;
     const MAX_BOOT2_INPUT_LEN: usize = BOOT2_OUTPUT_LEN - 4;
     let input_path: &Path = input_path.as_ref();
@@ -89,7 +88,8 @@ fn make_padded_bin<P: AsRef<Path>, Q: AsRef<Path>>(input_path: P, out_dir: Q) {
     let mut result_file = PathBuf::from(input_path.file_name().unwrap());
     result_file.set_extension("padded.bin");
     let result_path = out_dir.as_ref().join(result_file);
-    fs::write(result_path, blob).expect("writing padded output file");
+    fs::write(&result_path, blob).expect("writing padded output file");
+    result_path
 }
 
 #[cfg(feature = "assemble")]
@@ -100,24 +100,64 @@ fn calc_crc(data: &[u8]) -> u32 {
 }
 
 #[cfg(feature = "assemble")]
+fn update_precompiled_bin<P: AsRef<Path>>(input_path: P) {
+    let input_path: &Path = input_path.as_ref();
+    // Abort if this crate is being built as a dependency.
+    // This check is crude, but CARGO_PRIMARY_PACKAGE is not
+    // available in build scripts.
+    if !env::var("OUT_DIR")
+        .unwrap()
+        .starts_with(&env::var("CARGO_MANIFEST_DIR").unwrap())
+    {
+        panic!(
+            "UPDATE_PRECOMPILED_BINARIES must only be used when compiling this package directly"
+        );
+    }
+    let precompiled_bin_dir = env::var("CARGO_MANIFEST_DIR").unwrap() + "/bin/";
+    std::fs::copy(
+        &input_path,
+        Path::new(&precompiled_bin_dir).join(input_path.file_name().unwrap()),
+    )
+    .unwrap();
+}
+
+#[cfg(feature = "assemble")]
 fn main() -> Result<(), String> {
-    // Store temporary build files here
     let out_dir = env::var("OUT_DIR").unwrap();
-    // And our final output here
-    let final_outdir = concat!(env!("CARGO_MANIFEST_DIR"), "/bin/");
     for asm_file in SOURCE_FILES.iter() {
         let elf = make_elf(asm_file, &out_dir);
         let bin = make_bin(elf, &out_dir);
-        let _padded_bin = make_padded_bin(bin, &final_outdir);
+        let padded_bin = make_padded_bin(bin, &out_dir);
+        if env::var("UPDATE_PRECOMPILED_BINARIES").is_ok() {
+            update_precompiled_bin(padded_bin);
+        }
         println!("cargo:rerun-if-changed={}", asm_file);
     }
     println!("cargo:rerun-if-changed=./build.rs");
+    println!("cargo:rerun-if-env-changed=UPDATE_PRECOMPILED_BINARIES");
 
     Ok(())
 }
 
 #[cfg(not(feature = "assemble"))]
 fn main() -> Result<(), String> {
+    let in_dir = env::var("CARGO_MANIFEST_DIR").unwrap() + "/bin/";
+    let out_dir = env::var("OUT_DIR").unwrap();
+
+    let paths: Vec<_> = fs::read_dir(in_dir)
+        .unwrap()
+        .map(|entry| entry.unwrap().path())
+        .collect();
+    for path in paths {
+        if path
+            .file_name()
+            .unwrap()
+            .to_string_lossy()
+            .ends_with(".padded.bin")
+        {
+            std::fs::copy(&path, Path::new(&out_dir).join(path.file_name().unwrap())).unwrap();
+        }
+    }
     println!("cargo:warning=Using prebuilt boot2 files. use feature `assemble` to rebuild instead (requires GNU toolchain)");
     Ok(())
 }

--- a/check-blobs.sh
+++ b/check-blobs.sh
@@ -18,7 +18,7 @@ for lib in bin/*.bin; do
     arm-none-eabi-objdump -b binary -m armv6-m -M force-thumb -D "$lib" > "bin/${filename%.bin}.before"
 done
 
-cargo build --features=assemble
+UPDATE_PRECOMPILED_BINARIES=true cargo build --features=assemble
 
 for lib in bin/*.bin; do
     filename=$(basename "$lib")

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -2,28 +2,28 @@
 
 /// The bootloader to use if you have a W25Q080 flash device
 pub static BOOT_LOADER_W25Q080: [u8; 256] =
-    *include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/bin/boot2_w25q080.padded.bin"));
+    *include_bytes!(concat!(env!("OUT_DIR"), "/boot2_w25q080.padded.bin"));
 
 /// The bootloader to use if you want to copy code to RAM and then boot from RAM
 pub static BOOT_LOADER_RAM_MEMCPY: [u8; 256] =
-    *include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/bin/boot2_ram_memcpy.padded.bin"));
+    *include_bytes!(concat!(env!("OUT_DIR"), "/boot2_ram_memcpy.padded.bin"));
 
 /// The bootloader to use if you want to boot from an AT25SF128A flash device
 pub static BOOT_LOADER_AT25SF128A: [u8; 256] =
-    *include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/bin/boot2_at25sf128a.padded.bin"));
+    *include_bytes!(concat!(env!("OUT_DIR"), "/boot2_at25sf128a.padded.bin"));
 
 /// The bootloader to use if you want to boot from an GD25Q64CS flash device
 pub static BOOT_LOADER_GD25Q64CS: [u8; 256] =
-    *include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/bin/boot2_gd25q64cs.padded.bin"));
+    *include_bytes!(concat!(env!("OUT_DIR"), "/boot2_gd25q64cs.padded.bin"));
 
 /// The bootloader to use if you want to boot from an W25X10CL flash device
 pub static BOOT_LOADER_W25X10CL: [u8; 256] =
-    *include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/bin/boot2_w25x10cl.padded.bin"));
+    *include_bytes!(concat!(env!("OUT_DIR"), "/boot2_w25x10cl.padded.bin"));
 
 /// The bootloader to use if you want to boot from a generic flash device
 pub static BOOT_LOADER_GENERIC_03H: [u8; 256] =
-    *include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/bin/boot2_generic_03h.padded.bin"));
+    *include_bytes!(concat!(env!("OUT_DIR"), "/boot2_generic_03h.padded.bin"));
 
 /// The bootloader to use if you want to boot from an IS25LP080 flash device
 pub static BOOT_LOADER_IS25LP080: [u8; 256] =
-    *include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/bin/boot2_is25lp080.padded.bin"));
+    *include_bytes!(concat!(env!("OUT_DIR"), "/boot2_is25lp080.padded.bin"));

